### PR TITLE
ARROW-11565: [C++][Gandiva] Modify upper()/lower() logic to make them work for utf8 strings

### PR DESCRIPTION
**Pull request by sgnkc**
upper()/lower() in Gandiva used to work only for ASCII characters previously. This patch modifies the logic to make them work for all utf8 strings.

### CODE REVIEW

File: cpp/src/gandiva/precompiled/string_ops.cc [hunk 1]
```diff
@@ -219,28 +225,74 @@ const char* upper_utf8(gdv_int64 context, const char* data, gdv_int32 data_len,
...
+      } else {
+        out[out_idx++] = cur;
+      }
+      continue;
+    }
+
+    // Control reaches here when we encounter a multibyte character
+    gdv_uint8* in_char =
+        reinterpret_cast<gdv_uint8*>(gdv_fn_context_arena_malloc(context, char_len));
+    memcpy(in_char, data + i, char_len);
```
Comment (pprudhvi) [inline cpp/src/gandiva/precompiled/string_ops.cc:261]: can we avoid this by passing reference to data+i to UTF8Decode? similarily while encoding

Comment (sagnikc-dremio) [reply cpp/src/gandiva/precompiled/string_ops.cc]: Done.

File: cpp/src/gandiva/precompiled/string_ops.cc [hunk 2]
```diff
@@ -249,25 +297,73 @@ const char* lower_utf8(gdv_int64 context, const char* data, gdv_int32 data_len,
...
+        out[out_idx++] = static_cast<char>(cur + 0x20);
+      } else {
+        out[out_idx++] = cur;
+      }
+      continue;
+    }
+
+    // Control reaches here when we encounter a multibyte character
+    gdv_uint8* in_char =
+        reinterpret_cast<gdv_uint8*>(gdv_fn_context_arena_malloc(context, char_len));
```
Comment (pprudhvi) [inline cpp/src/gandiva/precompiled/string_ops.cc:332]: this can be avoided

File: cpp/src/gandiva/precompiled/string_ops.cc [hunk 3]
```diff
@@ -249,25 +297,73 @@ const char* lower_utf8(gdv_int64 context, const char* data, gdv_int32 data_len,
...
     }
-    ret[i] = cur;
+
+    // Convert the encoded codepoint to its lowercase codepoint
+    gdv_int32 lower_codepoint = utf8proc_tolower(char_codepoint);
+
+    // UTF8Encode advances the pointer by the number of bytes present in the lowercase
+    // character
+    gdv_uint8* out_char =
+        reinterpret_cast<gdv_uint8*>(gdv_fn_context_arena_malloc(context, 4));
```
Comment (pprudhvi) [inline cpp/src/gandiva/precompiled/string_ops.cc:352]: this too

File: cpp/src/gandiva/precompiled/string_ops.cc (hunk 2 above)
Comment (sagnikc-dremio) [reply cpp/src/gandiva/precompiled/string_ops.cc]: Yeah I'm sorry, I missed that one. Made the changes.

**Comment by sagnikc-dremio** [comment]
@maartenbreddels @kou

Hi Maarten/Sutou, I am facing some difficulties with respect to adding the utf8proc dependency for the case conversion of utf8 strings. Can you please review this PR and let me know what changes need to be made to address the above build failures?

**Comment by kou** [comment]
Could you try the following patch?

```diff
diff --git a/cpp/cmake_modules/DefineOptions.cmake b/cpp/cmake_modules/DefineOptions.cmake
index e4df40d61..c0abea7a2 100644
--- a/cpp/cmake_modules/DefineOptions.cmake
+++ b/cpp/cmake_modules/DefineOptions.cmake
@@ -363,7 +363,7 @@ if("${CMAKE_SOURCE_DIR}" STREQUAL "${CMAKE_CURRENT_SOURCE_DIR}")
 
   define_option(
     ARROW_WITH_UTF8PROC
-    "Build with support for Unicode properties using the utf8proc library;(only used if ARROW_COMPUTE is ON)"
+    "Build with support for Unicode properties using the utf8proc library;(only used if ARROW_COMPUTE is ON or ARROW_GANDIVA is ON)"
     ON)
   define_option(
     ARROW_WITH_RE2
diff --git a/cpp/src/gandiva/precompiled/CMakeLists.txt b/cpp/src/gandiva/precompiled/CMakeLists.txt
index 7343bc052..bc87c5ee6 100644
--- a/cpp/src/gandiva/precompiled/CMakeLists.txt
+++ b/cpp/src/gandiva/precompiled/CMakeLists.txt
@@ -77,6 +77,9 @@ foreach(SRC_FILE ${PRECOMPILED_SRCS})
            ${ARROW_GANDIVA_PC_CXX_FLAGS}
            -I${CMAKE_SOURCE_DIR}/src
            -I${ARROW_BINARY_DIR}/src)
+  if(ARROW_WITH_UTF8PROC)
+    list(APPEND PRECOMPILE_COMMAND -DARROW_WITH_UTF8PROC)
+  endif()
   if(NOT ARROW_USE_NATIVE_INT128)
     list(APPEND PRECOMPILE_COMMAND -I${Boost_INCLUDE_DIR})
   endif()
```

**Comment by sagnikc-dremio** [comment]
> Could you try the following patch?
> 
> ```diff
> diff --git a/cpp/cmake_modules/DefineOptions.cmake b/cpp/cmake_modules/DefineOptions.cmake
> index e4df40d61..c0abea7a2 100644
> --- a/cpp/cmake_modules/DefineOptions.cmake
> +++ b/cpp/cmake_modules/DefineOptions.cmake
> @@ -363,7 +363,7 @@ if("${CMAKE_SOURCE_DIR}" STREQUAL "${CMAKE_CURRENT_SOURCE_DIR}")
>  
>    define_option(
>      ARROW_WITH_UTF8PROC
> -    "Build with support for Unicode properties using the utf8proc library;(only used if ARROW_COMPUTE is ON)"
> +    "Build with support for Unicode properties using the utf8proc library;(only used if ARROW_COMPUTE is ON or ARROW_GANDIVA is ON)"
>      ON)
>    define_option(
>      ARROW_WITH_RE2
> diff --git a/cpp/src/gandiva/precompiled/CMakeLists.txt b/cpp/src/gandiva/precompiled/CMakeLists.txt
> index 7343bc052..bc87c5ee6 100644
> --- a/cpp/src/gandiva/precompiled/CMakeLists.txt
> +++ b/cpp/src/gandiva/precompiled/CMakeLists.txt
> @@ -77,6 +77,9 @@ foreach(SRC_FILE ${PRECOMPILED_SRCS})
>             ${ARROW_GANDIVA_PC_CXX_FLAGS}
>             -I${CMAKE_SOURCE_DIR}/src
>             -I${ARROW_BINARY_DIR}/src)
> +  if(ARROW_WITH_UTF8PROC)
> +    list(APPEND PRECOMPILE_COMMAND -DARROW_WITH_UTF8PROC)
> +  endif()
>    if(NOT ARROW_USE_NATIVE_INT128)
>      list(APPEND PRECOMPILE_COMMAND -I${Boost_INCLUDE_DIR})
>    endif()
> ```

@kou The above patch does not seem to work. After adding the option in precompiled/CMakeLists.txt, compilation fails with:

`fatal error: 'utf8proc.h' file not found`

**Comment by kou** [comment]
How about this?

```diff
diff --git a/cpp/cmake_modules/DefineOptions.cmake b/cpp/cmake_modules/DefineOptions.cmake
index e4df40d61..c0abea7a2 100644
--- a/cpp/cmake_modules/DefineOptions.cmake
+++ b/cpp/cmake_modules/DefineOptions.cmake
@@ -363,7 +363,7 @@ if("${CMAKE_SOURCE_DIR}" STREQUAL "${CMAKE_CURRENT_SOURCE_DIR}")
 
   define_option(
     ARROW_WITH_UTF8PROC
-    "Build with support for Unicode properties using the utf8proc library;(only used if ARROW_COMPUTE is ON)"
+    "Build with support for Unicode properties using the utf8proc library;(only used if ARROW_COMPUTE is ON or ARROW_GANDIVA is ON)"
     ON)
   define_option(
     ARROW_WITH_RE2
diff --git a/cpp/src/gandiva/precompiled/CMakeLists.txt b/cpp/src/gandiva/precompiled/CMakeLists.txt
index 7343bc052..76148a3c7 100644
--- a/cpp/src/gandiva/precompiled/CMakeLists.txt
+++ b/cpp/src/gandiva/precompiled/CMakeLists.txt
@@ -77,6 +77,16 @@ foreach(SRC_FILE ${PRECOMPILED_SRCS})
            ${ARROW_GANDIVA_PC_CXX_FLAGS}
            -I${CMAKE_SOURCE_DIR}/src
            -I${ARROW_BINARY_DIR}/src)
+  if(ARROW_WITH_UTF8PROC)
+    list(APPEND PRECOMPILE_COMMAND -DARROW_WITH_UTF8PROC)
+    if(UTF8PROC_COMPILER_DEFINITIONS)
+      # This doesn't work when multiple definitions exist.
+      list(APPEND PRECOMPILE_COMMAND "-D${UTF8PROC_COMPILER_DEFINITIONS}")
+    endif()
+    if(UTF8PROC_INCLUDE_DIR)
+      list(APPEND PRECOMPILE_COMMAND "-I${UTF8PROC_INCLUDE_DIR}")
+    endif()
+  endif()
   if(NOT ARROW_USE_NATIVE_INT128)
     list(APPEND PRECOMPILE_COMMAND -I${Boost_INCLUDE_DIR})
   endif()
```

**Comment by sagnikc-dremio** [comment]
@kou This patch seems to resolve the dependency issue. Thanks again.

**Comment by li-wu** [comment]
Any update for the PR?

**Comment by projjal** [comment]
@sagnikc-dremio The failing test might be because the generated jit code can't resolve the symbols corresponding to the utf8 library. Can you add the function in gdv_fn_stubs and check.

**Comment by anthonylouisbsb** [comment]
@projjal @praveenbingo I think this PR can be closed, since [the one](https://github.com/apache/arrow/pull/10040) I continued the implementation was merged.

**Pull request closed without merging**